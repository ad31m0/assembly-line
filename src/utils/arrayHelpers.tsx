/* istanbul ignore file */
/**
 *  Wrapper function for array push (append)
 * 
 * @param stage: string []
 * @param item: string
 * @returns stage: string[]
 */
export const push = (stage: string[], item: string) => {
    stage.push(item);
    return stage;
}
/**
 *  Wrapper function for array unshift (prepend)
 * 
 * @param stage: string []
 * @param item: string
 * @returns stage: string[]
 */
export const unshift = (stage: string[], item: string) => {
    stage.unshift(item);
    return stage;
}
/**
 *  Wrapper function for array remove with given item index
 * 
 * @param stage: string []
 * @param itemIdx: string
 * @returns stage: string[]
 */
export const remove = (stage: string[], itemIdx: number) => {
    stage.splice(itemIdx, 1);
    return stage;
}