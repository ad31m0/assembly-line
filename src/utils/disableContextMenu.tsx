import { MouseEvent } from "react";
/**
 * Disables default context menu by preventing the default handler, stopping propagation and returning false
 *
 * @param {MouseEvent} event
 * @public 
 */
export const disableContextMenu = (event: MouseEvent) => {
  if (event.preventDefault !== undefined)
    event.preventDefault();
  if (event.stopPropagation !== undefined)
    event.stopPropagation();
  return false;
};