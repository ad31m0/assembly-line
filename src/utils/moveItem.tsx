import { push, remove, unshift } from "./arrayHelpers";

export type moveItemType = (itemIdx: number, stageIdx: number, delta: number, itemName?: string) => void;

/**
 * Constructor for the moveItem function
 * shares read/edit access from an arbitrary context for testing convenience
 *   
 * @param stagesItems: string[][] 
 * @param setStagesItems: (items: string[][]) => void 
 * 
 * @returns (itemIdx: number, stageIdx: number, delta: number, itemName?: string) => void

 */
export const moveItemCtor = ({ stagesItems, setStagesItems }: { stagesItems: string[][], setStagesItems: (items: string[][]) => void }) => {
    const moveItem: moveItemType = (itemIdx, stageIdx, delta, itemName) => {
        // if valid stageIdx
        if (stageIdx > -1 && stageIdx < stagesItems.length) {
            // retrieve item name and remove from current stage
            itemName = stagesItems[stageIdx][itemIdx];
            stagesItems[stageIdx] = remove(stagesItems[stageIdx], itemIdx);
        }
        // if not left clicked at the [first] stage and not right clicked at the [last] stage
        if (!(stageIdx === 0 && delta < 0) && !(stageIdx === stagesItems.length - 1 && delta > 0)) {
            // get the destination stage based on the given delta (+1 || -1)
            const stage = stagesItems[stageIdx + delta];
            if (delta > 0) {
                // prepend if moving forward
                stagesItems[stageIdx + delta] = unshift(stage, itemName as string);
            } else {
                // append if moving backwards
                stagesItems[stageIdx + delta] = push(stage, itemName as string);
            }
        }
        setStagesItems([...stagesItems]);
    }
    return moveItem;
}