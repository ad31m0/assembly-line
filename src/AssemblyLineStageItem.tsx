import { moveItemType } from "./utils/moveItem";
/**
 *  a component for the each stage in the assembly line
 *  Should render the stage name and the list of stageItems provided
 * 
 * @param itemIdx: number index of the current item in stage
 * @param itemName: string name of the current item for display
 * @param stageIdx: number index of the current stage, used to send moveitem params
 * @param moveItem: (itemIdx: number, stageIdx: number, delta: number, itemName?: string) => void
 */
export const AssemblyLineStageItem = ({ itemIdx, itemName, stageIdx, moveItem }: { itemIdx: number, itemName: string, stageIdx: number, moveItem: moveItemType }) => {
  return (
    <div key={`${stageIdx}_${itemIdx}`}>
      <button
        className={`animate__animated animate__fadeIn`}
        onMouseDown={(event) => {
          switch (event.button) {
            case 0:
              // if left click move the item forward by passing delta = +1
              moveItem(itemIdx, stageIdx, +1);
              break;
            case 2:
              // if right click move the item backwards by passing delta = -1
              moveItem(itemIdx, stageIdx, -1);
              break;
            default:
              // unsupported do nothing;
              break;
          }
        }}>
        {itemName}
      </button>
    </div>
  );
}