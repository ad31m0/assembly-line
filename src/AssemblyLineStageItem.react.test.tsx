import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { configure, shallow } from 'enzyme';
import { AssemblyLineStageItem } from './AssemblyLineStageItem';
import { moveItemType } from './utils/moveItem';


configure({ adapter: new Adapter() });

describe('<AssemblyLineStageItem />', () => {
  const moveItemMock = jest.fn() as jest.MockedFunction<moveItemType>;
  moveItemMock.mockImplementation(() => { });
  const container = shallow(<AssemblyLineStageItem {...{ itemIdx: 0, itemName: "item1", stageIdx: 0, moveItem: moveItemMock }} />);
  it('should match the snapshot', () => {
    expect(container.html()).toMatchSnapshot();
  });

  it('should not move an item on any unsupported mouse click', () => {
    container.find('button').simulate('mouseDown', { button: 1 });
    expect(moveItemMock).not.toHaveBeenCalled();
  });

  it('should move item to next stage on left click', () => {
    container.find('button').simulate('mouseDown', { button: 0 });
    expect(moveItemMock).toBeCalledWith(0, 0, +1);
  });

  it('should move item to previous stage on right click', () => {
    container.find('button').simulate('mouseDown', { button: 2 });
    expect(moveItemMock).toBeCalledWith(0, 0, -1);
  });
});
