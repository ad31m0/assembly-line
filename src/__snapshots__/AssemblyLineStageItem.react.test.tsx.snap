// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AssemblyLineStageItem /> should match the snapshot 1`] = `"<div><button class=\\"animate__animated animate__fadeIn\\">item1</button></div>"`;
